### PR TITLE
LPD-27037 - Fix onAfterUpdate trigger to handle oldValue variable

### PR DESCRIPTION
**Pull request by G-Sa**
Related to :[LPD-27037](https://liferay.atlassian.net/browse/LPD-27037)

### CODE REVIEW

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java [hunk 1]
```diff
@@ -1545,6 +1545,8 @@ public ObjectEntry updateObjectEntry(
 
 		objectEntryPersistence.clearCache(SetUtil.fromArray(objectEntryId));
 
+		ObjectEntry originalObjectEntry = objectEntry;
+
 		objectEntry = objectEntryPersistence.findByPrimaryKey(objectEntryId);
```
Comment (carolmariaabb) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java:1550]: @G-Sa can you try to do something similar to this [code](https://github.com/carolmariaabb/liferay-portal/blob/344689f54b5ede48c9610059aed6ca63f777ba41/modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectFieldLocalServiceImpl.java#L1316-L1319)?

Comment (carolmariaabb) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java:1550]: @G-Sa we are calling `objectEntryPersistence.findByPrimaryKey` at the beginning of this method ([here](https://github.com/liferay-objects/liferay-portal/blob/8be3be73f0095c7256a43ca42e0eb16bc98a7181/modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java#L1512)). Why did you decide to save the original only at this point in the code?

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectActionLocalServiceTest.java [hunk 2]
```diff
@@ -992,8 +992,52 @@ public void testAddObjectActionWithConditionExpression() throws Exception {
...
+			TestPropsValues.getUserId(), objectEntry.getObjectEntryId(),
+			HashMapBuilder.<String, Serializable>put(
+				"firstName", "Peter"
+			).build(),
+			ServiceContextTestUtil.getServiceContext());
+
+		Assert.assertEquals(
+			objectEntriesCount + 1,
+			_objectEntryLocalService.getObjectEntriesCount(
+				0, _objectDefinition.getObjectDefinitionId()));
```
Comment (carolmariaabb) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectActionLocalServiceTest.java:1036]: @G-Sa in addition to checking the number of entries, could you also check that a new entry with `Matheus` was created?

**Review by carolmariaabb — CHANGES_REQUESTED** [review]

**Comment by carolmariaabb** [comment]
@G-Sa please move the ticket to In Review

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java (hunk 1 above)
Comment (G-Sa) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java:1550]: I just went as far down as i could simply to be nearest to the point where the variable will be used

Comment (carolmariaabb) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java:1550]: @G-Sa but is it possible that before this code the `objectEntry` was already updated? I did not understand why we are calling the `findByPrimaryKey` method twice.

Can you check how it worked before (when using the model listener)?

**Comment by G-Sa** [comment]
ci:test:sf

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectActionLocalServiceTest.java [hunk 3]
```diff
@@ -992,8 +992,76 @@ public void testAddObjectActionWithConditionExpression() throws Exception {
...
+			TestPropsValues.getUserId(), objectEntry.getObjectEntryId(),
+			HashMapBuilder.<String, Serializable>put(
+				"firstName", "Andrew"
+			).build(),
+			ServiceContextTestUtil.getServiceContext());
+
+		Assert.assertEquals(
+			objectEntriesCount,
+			_objectEntryLocalService.getObjectEntriesCount(
+				0, _objectDefinition.getObjectDefinitionId()));
```
Comment (carolmariaabb) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectActionLocalServiceTest.java:1036]: @G-Sa can we simplify by removing this code? 

- create a new entry
- update to Paulo won't trigger the object action
- update to something else (oldValue = Paulo) will trigger the object action

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectActionLocalServiceTest.java [hunk 4]
```diff
@@ -992,8 +992,76 @@ public void testAddObjectActionWithConditionExpression() throws Exception {
...
+				"objectDefinitionId", _objectDefinition.getObjectDefinitionId()
+			).put(
+				"predefinedValues",
+				JSONUtil.putAll(
+					JSONUtil.put(
+						"inputAsValue", true
+					).put(
+						"name", "firstName"
+					).put(
+						"value", "Matheus"
```
Comment (carolmariaabb) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectActionLocalServiceTest.java:1009]: @G-Sa can we make this random?

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectActionLocalServiceTest.java [hunk 5]
```diff
@@ -992,8 +992,76 @@ public void testAddObjectActionWithConditionExpression() throws Exception {
...
+					)
+				).toString()
+			).build(),
+			false);
+
+		objectEntry = _objectEntryLocalService.addObjectEntry(
+			TestPropsValues.getUserId(), 0,
+			_objectDefinition.getObjectDefinitionId(),
+			HashMapBuilder.<String, Serializable>put(
+				"firstName", "Andre"
```
Comment (carolmariaabb) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectActionLocalServiceTest.java:1019]: @G-Sa can we make this random?

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectActionLocalServiceTest.java [hunk 6]
```diff
@@ -992,8 +992,76 @@ public void testAddObjectActionWithConditionExpression() throws Exception {
...
+
+		Assert.assertEquals(
+			objectEntriesCount,
+			_objectEntryLocalService.getObjectEntriesCount(
+				0, _objectDefinition.getObjectDefinitionId()));
+
+		_objectEntryLocalService.updateObjectEntry(
+			TestPropsValues.getUserId(), objectEntry.getObjectEntryId(),
+			HashMapBuilder.<String, Serializable>put(
+				"firstName", "Peter"
```
Comment (carolmariaabb) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectActionLocalServiceTest.java:1053]: @G-Sa can we make this random?

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectActionLocalServiceTest.java [hunk 7]
```diff
@@ -992,8 +992,76 @@ public void testAddObjectActionWithConditionExpression() throws Exception {
...
+			).build(),
+			ServiceContextTestUtil.getServiceContext());
+
+		Assert.assertEquals(
+			objectEntriesCount + 1,
+			_objectEntryLocalService.getObjectEntriesCount(
+				0, _objectDefinition.getObjectDefinitionId()));
+
 		_objectActionLocalService.deleteObjectAction(objectAction2);
 		_objectActionLocalService.deleteObjectAction(objectAction3);
```
Comment (carolmariaabb) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectActionLocalServiceTest.java:1063]: @G-Sa can we delete the `objectAction2` and `objectAction3` before creating the `objectAction4`? Like we are doing for the `objectAction1`

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectActionLocalServiceTest.java
Comment (carolmariaabb) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectActionLocalServiceTest.java:1]: my suggestion:

```
_objectActionLocalService.deleteObjectAction(objectAction2);
_objectActionLocalService.deleteObjectAction(objectAction3);

ObjectAction objectAction4 = _addObjectAction(
	"oldValue('firstName') == 'Paulo'", RandomTestUtil.randomString(),
	ObjectActionExecutorConstants.KEY_ADD_OBJECT_ENTRY,
	ObjectActionTriggerConstants.KEY_ON_AFTER_UPDATE,
	UnicodePropertiesBuilder.put(
		"objectDefinitionId", _objectDefinition.getObjectDefinitionId()
	).put(
		"predefinedValues",
		JSONUtil.putAll(
			JSONUtil.put(
				"inputAsValue", true
			).put(
				"name", "firstName"
			).put(
				"value", RandomTestUtil.randomString()
			)
		).toString()
	).build(),
	false);

objectEntry = _objectEntryLocalService.addObjectEntry(
	TestPropsValues.getUserId(), 0,
	_objectDefinition.getObjectDefinitionId(),
	HashMapBuilder.<String, Serializable>put(
		"firstName", RandomTestUtil.randomString()
	).build(),
	ServiceContextTestUtil.getServiceContext());

int objectEntriesCount = _objectEntryLocalService.getObjectEntriesCount(
	0, _objectDefinition.getObjectDefinitionId());

_objectEntryLocalService.updateObjectEntry(
	TestPropsValues.getUserId(), objectEntry.getObjectEntryId(),
	HashMapBuilder.<String, Serializable>put(
		"firstName", "Paulo"
	).build(),
	ServiceContextTestUtil.getServiceContext());

Assert.assertEquals(
	objectEntriesCount,
	_objectEntryLocalService.getObjectEntriesCount(
		0, _objectDefinition.getObjectDefinitionId()));

_objectEntryLocalService.updateObjectEntry(
	TestPropsValues.getUserId(), objectEntry.getObjectEntryId(),
	HashMapBuilder.<String, Serializable>put(
		"firstName", RandomTestUtil.randomString()
	).build(),
	ServiceContextTestUtil.getServiceContext());

Assert.assertEquals(
	objectEntriesCount + 1,
	_objectEntryLocalService.getObjectEntriesCount(
		0, _objectDefinition.getObjectDefinitionId()));

_objectActionLocalService.deleteObjectAction(objectAction4);
```

**Review by carolmariaabb — CHANGES_REQUESTED** [review]

**Comment by G-Sa** [comment]
ci:test:sf

**Review by carolmariaabb — APPROVED** [review]

**Comment by carolmariaabb** [comment]
ci:test:object

**Comment by carolmariaabb** [comment]
ci:test:relevant

**Comment by G-Sa** [comment]
ci:test:sf

**Review by carolmariaabb — APPROVED** [review]

**Comment by carolmariaabb** [comment]
ci:test:object

**Comment by carolmariaabb** [comment]
ci:test:relevant

**Comment by marcelabc** [comment]
Let's wait for ci:test:object to finish :+1:

**Pull request closed without merging**